### PR TITLE
[HGI-284] Add firebase support for Google Analytics 4 Cloud Destination

### PR DESCRIPTION
**Pull request by varadarajan-tw**
This PR adds Firebase support to Google Analytics 4 destination.

[SDD](https://docs.google.com/document/d/1h_-KtROTfD8aI9HUvSzCUzwzFHx_CDAmaSatFJGrfnM/edit#heading=h.9leaiz3rqya6)

**Note**
The SDD says that DataStreamType field is mandatory. However, after conversations, we have decided not to make it mandatory. Refer [this ticket](https://segment.atlassian.net/browse/HGI-336) for background.

* Adds a new setting - firebaseAppId and two new fields to mappings: DataStreamType(optional) and app_instance_id(optional) for firebase parameters
* Adds testAuthentication validation - validates if one of firebaseAppId or measurementId is configured.
* Updates client_id and measurement_id in mapping to be optional params
* Updates unit tests

## Testing

[Test Results](https://docs.google.com/document/d/1BhUMnNY2jAupzbBpYzZHqd6vyFoXMQnS5LAWJoh06jc/edit?pli=1#heading=h.t3e6o7a65fps)

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/destination-actions/src/destinations/google-analytics-4/ga4-properties.ts [hunk 1]
```diff
@@ -111,6 +112,14 @@ export const client_id: InputField = {
   }
 }
 
+export const app_instance_id: InputField = {
+  label: 'Firebase App Instance ID',
+  description:
+    'Uniquely identifies a specific installation of a Firebase app. This value needs to be retrieved through the Firebase SDK. Required for mobile app streams.',
```
Comment (maryamsharif) [inline packages/destination-actions/src/destinations/google-analytics-4/ga4-properties.ts:118]:
```suggestion
    'Uniquely identifies a specific installation of a Firebase app. This value needs to be retrieved through the Firebase SDK. **Required for mobile app streams.**',
```

File: packages/destination-actions/src/destinations/google-analytics-4/ga4-properties.ts [hunk 2]
```diff
@@ -101,7 +102,7 @@ export const client_id: InputField = {
   label: 'Client ID',
   description: 'Uniquely identifies a user instance of a web client.',
```
Comment (maryamsharif) [inline packages/destination-actions/src/destinations/google-analytics-4/ga4-properties.ts:103]:
```suggestion
  description: 'Uniquely identifies a user instance of a web client. **Required for web streams.**',
```

File: packages/destination-actions/src/destinations/google-analytics-4/index.ts [hunk 3]
```diff
@@ -32,9 +33,16 @@ const destination: DestinationDefinition<Settings> = {
       measurementId: {
         label: 'Measurement ID',
         description:
-          'The measurement ID associated with a stream. Found in the Google Analytics UI under: Admin > Data Streams > choose your stream > Measurement ID',
+          'The Measurement ID associated with a stream. Required for web streams. Found in the Google Analytics UI under: Admin > Data Streams > choose your stream > Measurement ID.',
```
Comment (maryamsharif) [inline packages/destination-actions/src/destinations/google-analytics-4/index.ts:36]:
```suggestion
          'The Measurement ID associated with a stream. Found in the Google Analytics UI under: Admin > Data Streams > choose your stream > Measurement ID. **Required for web streams. **',
```

File: packages/destination-actions/src/destinations/google-analytics-4/__tests__/addPaymentInfo.test.ts [hunk 4]
```diff
@@ -925,10 +955,135 @@ describe('GA4', () => {
...
+          event: testEvent,
+          settings: {
+            apiSecret
+          },
+          useDefaultMappings: true
+        })
+      ).rejects.toThrowError('Measurement ID is required for web streams')
+    })
+
+    it('should throw error when data stream type is web and client_is is not provided', async () => {
```
Comment (maryamsharif) [inline packages/destination-actions/src/destinations/google-analytics-4/__tests__/addPaymentInfo.test.ts:1071]:
```suggestion
    it('should throw error when data stream type is web and client_id is not provided', async () => {
```

Comment (maryamsharif) [reply packages/destination-actions/src/destinations/google-analytics-4/__tests__/addPaymentInfo.test.ts:1071]: I think this typo is in all the test cases, so just double check to make sure.

File: packages/destination-actions/src/destinations/google-analytics-4/__tests__/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { createTestIntegration } from '@segment/actions-core'
+import Definition from '../index'
+
+const testDestination = createTestIntegration(Definition)
+
+describe('Google Analytics 4', () => {
+  describe('testAuthentication', () => {
+    it('should not throw error when Measurement ID is provided but Firebase App ID is provided', async () => {
```
Comment (maryamsharif) [inline packages/destination-actions/src/destinations/google-analytics-4/__tests__/index.test.ts:8]:
```suggestion
    it('should not throw error when Measurement ID is provided but Firebase App ID is not provided', async () => {
```

File: packages/destination-actions/src/destinations/google-analytics-4/__tests__/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+      await expect(
+        testDestination.testAuthentication({ apiSecret: 'apiSecret', measurementId: 'G-M12312' })
+      ).resolves.not.toThrowError()
+    })
+    it('should not throw error when Firebase App ID is provided but Measurement ID is provided', async () => {
```
Comment (maryamsharif) [inline packages/destination-actions/src/destinations/google-analytics-4/__tests__/index.test.ts:13]:
```suggestion
    it('should not throw error when Firebase App ID is provided but Measurement ID is not provided', async () => {
```

File: packages/destination-actions/src/destinations/google-analytics-4/index.ts [hunk 7]
```diff
@@ -32,9 +33,16 @@ const destination: DestinationDefinition<Settings> = {
...
         type: 'string',
-        required: true
+        required: false
+      },
+      firebaseAppId: {
+        label: 'Firebase App ID',
+        description:
+          'The Firebase App ID associated with the Firebase app. Required for mobile app streams. Found in the Firebase console under: Project Settings > General > Your Apps > App ID.',
```
Comment (maryamsharif) [inline packages/destination-actions/src/destinations/google-analytics-4/index.ts:43]:
```suggestion
          'The Firebase App ID associated with the Firebase app. Found in the Firebase console under: Project Settings > General > Your Apps > App ID. **Required for mobile app streams.** ' ,
```

**Review by maryamsharif — COMMENTED** [review]
Amazing work! Just left a few comments regarding the descriptions. Excellent testing done as well! 🔥

File: packages/destination-actions/src/destinations/google-analytics-4/ga4-properties.ts (hunk 1 above)
Comment (varadarajan-tw) [reply packages/destination-actions/src/destinations/google-analytics-4/ga4-properties.ts]: Updated the recommended changes to description.

File: packages/destination-actions/src/destinations/google-analytics-4/ga4-properties.ts (hunk 2 above)
Comment (varadarajan-tw) [reply packages/destination-actions/src/destinations/google-analytics-4/ga4-properties.ts]: Updated the recommended changes to description.

**Comment by varadarajan-tw** [comment]
Made DataStreamType as an optional field to reduce Risk. Refer [JIRA Ticket](https://segment.atlassian.net/browse/HGI-336)
Updated testing result for DataStreamType defaulting to *Web* if no value is set. Refer  [here](https://segment.atlassian.net/browse/HGI-336).

**Review by kdaswani — COMMENTED** [review]
I reviewed the new fields and descriptions, and all looks great! I will defer to eng to do the final code review. Happy to test whenever available in my staging workspace (`kiara-test`). Thanks @varadarajan-tw !

**Review by maryamsharif — APPROVED** [review]

File: packages/destination-actions/src/destinations/google-analytics-4/viewPromotion/index.ts [hunk 8]
```diff
@@ -113,7 +130,8 @@ const action: ActionDefinition<Settings, Payload> = {
       request_object.timestamp_micros = convertTimestamp(payload.timestamp_micros)
     }
 
-    return request('https://www.google-analytics.com/mp/collect', {
+    // Firebase App ID can contain colons(:) and they should not be encoded. Hence, interpolating search params to url string instead of passing them as search_params
+    return request(`https://www.google-analytics.com/mp/collect?${stream_params.search_params}`, {
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/google-analytics-4/viewPromotion/index.ts:134]: NIT: I wonder if it makes sense to move this into a shared file as it's on every updated action. A subsequent update will require the maintainer to go into multiple files to perform the same change.

Comment (varadarajan-tw) [reply packages/destination-actions/src/destinations/google-analytics-4/viewPromotion/index.ts]: Makes sense. Here is the diff https://github.com/segmentio/action-destinations/pull/967/commits/efd898336f9c93a9e8310258dbdbc08f7e537234 with the change.

**Pull request merged**